File: applications/spring-geode-showcase/src/main/java/com/vmware/spring/geode/showcase/controller/AccountController.java
package com.vmware.spring.geode.showcase.controller;

import com.vmware.spring.geode.showcase.domain.Account;
import com.vmware.spring.geode.showcase.repostories.AccountRepository;
import lombok.AllArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@AllArgsConstructor
@RestController
public class AccountController
{
    private final AccountRepository accountRepository;

    @PostMapping("save")
    public <S extends Account> S save(@RequestBody S s)
    {
        return accountRepository.save(s);
    }

    @GetMapping("findById")
    public Optional<Account> findById(String s)
    {
        return accountRepository.findById(s);
    }

    @DeleteMapping("deleteById/{id}")
    public void deleteById(@PathVariable String id)
    {
        accountRepository.deleteById(id);
    }
}

File: applications/account/src/main/kotlin/com/vmware/financial/open/banking/controller/AccountController.kt
package com.vmware.financial.open.banking.controller

import com.vmware.financial.open.banking.domain.AccountLocation
import com.vmware.financial.open.banking.repository.AccountRepository
import com.vmware.financial.open.banking.repository.LocationRepository
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import java.lang.IllegalArgumentException

/**
 * @author Gregory Green
 */
@RestController
class AccountController(
    private val accountRepository: AccountRepository,
    private val locationRepository: LocationRepository
) {
    private val validZipRegEx = "^\\d{5}(?:[-\\s]\\d{4})?\$".toRegex();

    @PostMapping("save")
    @Transactional
    fun save(@RequestBody accountLocation: AccountLocation) {
        accountRepository.save(accountLocation.account)

        var location = accountLocation.location;
        if(!location.zipCode.matches(validZipRegEx))
            throw IllegalArgumentException("Invalid zip code ${location.zipCode}");

        locationRepository.save(accountLocation.location)
    }

}